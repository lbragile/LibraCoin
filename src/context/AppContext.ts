import React from "react";
import { IAction, IState } from "../typings/AppTypes";

type TAppContext = { state: IState; dispatch: React.Dispatch<IAction> } | undefined;

export const AppContext = React.createContext<TAppContext>(undefined);
