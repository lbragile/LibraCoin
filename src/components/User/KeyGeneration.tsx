import React, { useRef, useState, useContext, useEffect } from "react";
import { Form, InputGroup } from "react-bootstrap";

import { AppContext } from "../../context/AppContext";
import { IAction, IState } from "../../typings/AppTypes";
import { ACTIONS } from "../../enums/AppDispatchActions";
import { copyKey } from "../../utils/copyInput";
import { CryptoKeyToHex } from "../../utils/conversion";

import "./User.scss";

export default function KeyGeneration(): JSX.Element {
  const { state, dispatch } = useContext(AppContext) as { state: IState; dispatch: React.Dispatch<IAction> };

  const numRows = useRef(4);
  const publicKeyRef = useRef<HTMLTextAreaElement>(null);
  const privateKeyRef = useRef<HTMLTextAreaElement>(null);

  const [copied, setCopied] = useState<boolean[]>([false, false]);

  // add a user if it's the first time visiting
  useEffect(() => {
    async function addUser(): Promise<void> {
      if (!state.user.publicKey) {
        const { publicKey, privateKey } = await window.crypto.subtle.generateKey(
          { name: "ECDSA", namedCurve: "P-256" },
          true,
          ["sign", "verify"]
        );

        const publicKeyStr = await CryptoKeyToHex("spki", publicKey as CryptoKey);
        const privateKeyStr = await CryptoKeyToHex("pkcs8", privateKey as CryptoKey);
        if (publicKeyRef.current && privateKeyRef.current) {
          publicKeyRef.current.innerText = publicKeyStr;
          privateKeyRef.current.innerText = new Array(privateKeyStr.length).fill("â—¦").join("");
        }

        const balance = Number(1000).toFixed(2);
        const mainUser = { publicKey: publicKeyStr, privateKey: privateKeyStr, balance };
        dispatch({ type: ACTIONS.SET_MAIN_USER, payload: { user: mainUser } });

        const newUsers = [...state.users, { publicKey: publicKeyStr, balance }];
        dispatch({ type: ACTIONS.UPDATE_USERS, payload: { users: newUsers } });
      }
    }

    addUser();
  }, [dispatch, state.user.publicKey, state.users]);

  const togglePrivateKey = () => {
    const show = privateKeyRef.current?.value.includes("â—¦");
    const hiddenVal = new Array(state.user.privateKey.length).fill("â—¦").join("");
    (privateKeyRef.current as HTMLTextAreaElement).value = show ? state.user.privateKey : hiddenVal;
  };

  return (
    <div className="container-fluid d-flex justify-content-center mx-auto row my-5">
      <InputGroup className="user-key col-12 col-lg-5 pl-3 pl-lg-0">
        <InputGroup.Prepend>
          <InputGroup.Text>Public</InputGroup.Text>
        </InputGroup.Prepend>

        <Form.Control
          aria-label="publicKey"
          as="textarea"
          rows={numRows.current}
          className="rounded-right"
          defaultValue={state.user?.publicKey ?? ""}
          isValid={copied[0]}
          onFocus={(e: React.FocusEvent<HTMLTextAreaElement>) => copyKey(e, setCopied, "public")}
          onBlur={() => setCopied([false, false])}
          readOnly
          ref={publicKeyRef}
        />

        <Form.Control.Feedback type="valid">Copied to clipboard!</Form.Control.Feedback>
      </InputGroup>

      <InputGroup className="user-key col-12 col-lg-5 pl-3">
        <InputGroup.Prepend>
          <InputGroup.Text>Private</InputGroup.Text>
        </InputGroup.Prepend>

        <Form.Control
          aria-label="privateKey"
          as="textarea"
          rows={numRows.current}
          defaultValue={state.user?.privateKey ? new Array(state.user.privateKey.length).fill("â—¦").join("") : ""}
          onFocus={(e: React.FocusEvent<HTMLTextAreaElement>) => copyKey(e, setCopied, "private")}
          onBlur={() => setCopied([false, false])}
          isValid={copied[1]}
          readOnly
          ref={privateKeyRef}
        />

        <InputGroup.Append>
          <InputGroup.Text className="rounded-right">
            <span id="private-reveal-eyes" onClick={togglePrivateKey}>
              ðŸ‘€
            </span>
          </InputGroup.Text>
        </InputGroup.Append>

        <Form.Control.Feedback type="valid">Copied to clipboard!</Form.Control.Feedback>
      </InputGroup>
    </div>
  );
}
