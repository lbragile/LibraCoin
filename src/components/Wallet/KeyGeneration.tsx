import React, { useRef, useState, useContext } from "react";
import { Button, Form } from "react-bootstrap";

import { Wallet } from "../Wallet/Wallet_class";
import { AppContext } from "../../context/AppContext";
import { IAction, IState } from "../../typings/AppTypes";
import { ACTIONS } from "../../enums/AppDispatchActions";
import { copyKey } from "../../utils/copyInput";

import "./Wallet.css";

export default function KeyGeneration(): JSX.Element {
  const { state, dispatch } = useContext(AppContext) as { state: IState; dispatch: React.Dispatch<IAction> };

  const publicKey = useRef<HTMLTextAreaElement>(null);
  const privateKey = useRef<HTMLTextAreaElement>(null);

  const numRows = useRef(4);

  const [copied, setCopied] = useState<boolean[]>([false, false]);

  const addUser = async () => {
    const user = new Wallet();
    await user.initialize();

    const publicKeyStr = await Wallet.CryptoKeyToHex("spki", user.publicKey);
    const privateKeyStr = await Wallet.CryptoKeyToHex("pkcs8", user.privateKey);
    if (publicKey.current && privateKey.current) {
      publicKey.current.innerText = publicKeyStr;
      privateKey.current.innerText = new Array(privateKeyStr.length).fill("â—¦").join("");
    }

    const newUsers = [...state.users, { publicKey: publicKeyStr, balance: user.balance }];
    localStorage.setItem("user", JSON.stringify({ publicKey: publicKeyStr, privateKey: privateKeyStr, balance: user.balance })); // prettier-ignore
    localStorage.setItem("users", JSON.stringify(newUsers));
    dispatch({ type: ACTIONS.UPDATE_USERS, payload: { users: newUsers } });
  };

  const togglePrivateKey = () => {
    if (privateKey.current) {
      if (privateKey.current.value.includes("â—¦")) {
        privateKey.current.value = JSON.parse(localStorage.getItem("user") as string).privateKey;
      } else {
        privateKey.current.value = new Array(privateKey.current.value.length).fill("â—¦").join("");
      }
    }
  };

  return (
    <div className="row">
      {!JSON.parse(localStorage.getItem("user") as string)?.publicKey ? (
        <div className="col-2">
          <Button variant="primary" className="p-3 font-weight-bold" onClick={addUser}>
            Create Wallet
          </Button>
        </div>
      ) : (
        <div className="col-1"></div>
      )}
      <div className="col-5">
        <Form.Group className="user-key">
          <Form.Label className="mb-3">
            <h3>Public Key:</h3>
          </Form.Label>
          <Form.Control
            as="textarea"
            rows={numRows.current}
            defaultValue={localStorage.getItem("user") && JSON.parse(localStorage.getItem("user") as string).publicKey}
            isValid={copied[0]}
            onFocus={(e: React.FocusEvent<HTMLTextAreaElement>) => copyKey(e, setCopied, "public")}
            ref={publicKey}
          />
          <Form.Control.Feedback type="valid">Copied to clipboard!</Form.Control.Feedback>
        </Form.Group>
      </div>
      <div className="col-5">
        <Form.Group className="user-key">
          <Form.Label className="mb-3">
            <h3>
              Private Key:{" "}
              {JSON.parse(localStorage.getItem("user") as string)?.publicKey && (
                <span id="private-reveal-eyes" onClick={togglePrivateKey}>
                  ðŸ‘€
                </span>
              )}
            </h3>
          </Form.Label>
          <Form.Control
            as="textarea"
            rows={numRows.current}
            defaultValue={
              localStorage.getItem("user")
                ? new Array(JSON.parse(localStorage.getItem("user") as string).privateKey.length).fill("â—¦").join("")
                : ""
            }
            onFocus={(e: React.FocusEvent<HTMLTextAreaElement>) => copyKey(e, setCopied, "private")}
            isValid={copied[1]}
            ref={privateKey}
          />
          <Form.Control.Feedback type="valid">Copied to clipboard!</Form.Control.Feedback>
        </Form.Group>
      </div>
    </div>
  );
}
