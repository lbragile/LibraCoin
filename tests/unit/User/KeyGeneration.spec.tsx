import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";

import KeyGeneration from "../../../src/components/User/KeyGeneration";
import { AppContext } from "../../../src/context/AppContext";

const { state, dispatch } = global;

it("renders correctly", () => {
  const { getByLabelText, getByRole } = render(
    <AppContext.Provider value={{ state, dispatch }}>
      <KeyGeneration />
    </AppContext.Provider>
  );

  const publicTitle = getByLabelText("Public:");
  const privateTitle = getByLabelText("Private:");
  const publicField = screen.getByText("Private:", { selector: "heading" });
  const privateField = screen.getByText("Public:", { selector: "heading" });

  [publicTitle, privateTitle, publicField, privateField].forEach((elem) => {
    expect(elem).toBeInTheDocument();
  });

  expect(screen).toMatchSnapshot();
});

// describe("input field text", () => {
//   const privateKey = "308187020100301306072a8648ce3d020106082a8648ce3";

//   it("has no user in localStorage", () => {
//     localStorage.removeItem("user");

//     const { container } = render(
//       <AppContext.Provider value={{ state, dispatch }}>
//         <KeyGeneration />
//       </AppContext.Provider>
//     );

//     const publicField = container.querySelector("#publicKey") as HTMLTextAreaElement;
//     const privateField = container.querySelector("#privateKey") as HTMLTextAreaElement;
//     const eyes = container.querySelector("#private-reveal-eyes");

//     expect(publicField.value).toBe("");
//     expect(privateField.value).toBe("");
//     expect(eyes).toBeFalsy();
//   });

//   it("has user in localStorage", () => {
//     localStorage.setItem("user", JSON.stringify({ ...state.users[0], privateKey }));

//     const { container } = render(
//       <AppContext.Provider value={{ state, dispatch }}>
//         <KeyGeneration />
//       </AppContext.Provider>
//     );

//     const publicField = container.querySelector("#publicKey") as HTMLTextAreaElement;
//     const privateField = container.querySelector("#privateKey") as HTMLTextAreaElement;
//     const eyes = container.querySelector("#private-reveal-eyes");

//     expect(publicField.value).toBe(state.users[0].publicKey);
//     expect(privateField.value).toBe(new Array(privateKey.length).fill("◦").join(""));
//     expect(eyes).toBeTruthy();
//   });

//   it("reveals private key on eye click", () => {
//     localStorage.setItem("user", JSON.stringify({ ...state.users[0], privateKey }));

//     const { container } = render(
//       <AppContext.Provider value={{ state, dispatch }}>
//         <KeyGeneration />
//       </AppContext.Provider>
//     );

//     const privateField = container.querySelector("#privateKey") as HTMLTextAreaElement;
//     const originalValue = privateField.value;
//     const eyes = container.querySelector("#private-reveal-eyes") as HTMLSpanElement;
//     fireEvent.click(eyes);

//     expect(originalValue).toBe(new Array(privateKey.length).fill("◦").join(""));
//     expect(privateField.value).toBe(privateKey);
//   });
// });
