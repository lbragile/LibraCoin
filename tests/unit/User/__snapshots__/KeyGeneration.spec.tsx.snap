// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<DocumentFragment>
  .c0 {
  margin: 0.25em 0.5em;
  word-wrap: break-word;
}

.c1:hover {
  cursor: pointer;
}

<div
    class="container col-11 col-lg-8 justify-content-center my-5 bg-dark rounded p-2"
  >
    <div
      class="c0 mx-auto input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="input-group-text"
        >
          Public Key (PK)
        </span>
      </div>
      <textarea
        aria-label="publicKey"
        class="rounded-right form-control"
        readonly=""
        rows="3"
      >
        3059301306072a8648ce3d020106082a8648ce3d030107034200048fa7d69599babb
      </textarea>
      <div
        class="valid-feedback"
      >
        Copied to clipboard!
      </div>
    </div>
    <div
      class="c0 mx-auto input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="input-group-text"
        >
          Private Key (SK)
        </span>
      </div>
      <textarea
        aria-label="privateKey"
        class="form-control"
        readonly=""
        rows="3"
      >
        ◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦◦
      </textarea>
      <div
        class="input-group-append"
      >
        <span
          class="rounded-right input-group-text"
        >
          <span
            aria-label="Key Reveal Eye"
            class="c1"
            role="img"
          >
            <svg
              fill="currentColor"
              height="1em"
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M10.5 8a2.5 2.5 0 1 1-5 0 2.5 2.5 0 0 1 5 0z"
              />
              <path
                d="M0 8s3-5.5 8-5.5S16 8 16 8s-3 5.5-8 5.5S0 8 0 8zm8 3.5a3.5 3.5 0 1 0 0-7 3.5 3.5 0 0 0 0 7z"
              />
            </svg>
             
          </span>
        </span>
      </div>
      <div
        class="valid-feedback"
      >
        Copied to clipboard!
      </div>
    </div>
  </div>
</DocumentFragment>
`;
